contracts/RockPaperScissors.sol

```diff
@@ -0,0 +1,286 @@
+        game.step = Steps.PLAYER_ONE_MOVE;
+        depositStake(stake);
+
+        emit PlayerOneMoves(msg.sender, gameKey, stake);
+    }
+
+    /*
+    * The opponent joins the game
+    */
+    function playerTwoMove(bytes32 gameKey, bytes32 moveHash) external payable whenNotPaused {
+
+        Game storage game = games[gameKey];
+        uint256 stake = game.stake;
+        require(stake != 0, GAME_NOT_FOUND_MSG);
+        require(game.step == Steps.PLAYER_ONE_MOVE, INVALID_STEP_MSG);
+
+        game.playerTwo = msg.sender;
+        game.playerTwoMoveHash = moveHash;
+        game.step = steps.PLAYER_TWO_MOVE;
+        depositStake(stake);
+
+        emit PlayerTwoMoves(msg.sender, msg.value, gameKey);
+    }
+
+    function depositStake(uint256 stake) internal {
+        uint256 senderBalance = balances[msg.sender];
+
+        if (msg.value == stake) {
+            return;
+        }
+
+        if (msg.value < stake) {
+            require(senderBalance >= stake.sub(msg.value), "Insufficient balance");
+            balances[msg.sender] = senderBalance.sub(stake.sub(msg.value));
+        } else {
+            balances[msg.sender] = senderBalance.add(msg.value.sub(stake));
+        }
+
+        Deposit(msg.sender, msg.value);
+    }
+
+    function playerOneReveal(bytes32 gameKey, bytes32 secret, uint8 move) external whenNotPaused {
+
+        require(move < 3, INVALID_MOVE_MSG);
+        Game storage game = games[gameKey];
+        uint256 stake = game.stake;
+        require(stake != 0, GAME_NOT_FOUND_MSG);
+        require(game.step == Steps.PLAYER_TWO_MOVE, INVALID_STEP_MSG);
+        address playerOne = game.playerOne;
+        require(msg.sender == playerOne, ONLY_PLAYER_ONE_CALL_MSG);
+
+
+        //Validate move
+        bytes32 expectedMoveHash = createGameKey(secret, move);
+        require(gameKey == expectedMoveHash, "Move and secret do not match");
+
+        game.playerOneMove = move;
+        game.step = Steps.PLAYER_ONE_REVEAL;
+    }
+
+    /**
+    * Once the opponent has made their move, the creator of the game must resolve to determine the winner
+    * and reward the winner.
+    * If the creator does not resolve the game then the opponent can collect their winnings
+     * after the forfeit window has expired.
+    */
+    function playerTwoReveal(bytes32 gameKey, bytes32 secret, uint8 playerTwoMove) external whenNotPaused {
+
+        Game storage game = games[gameKey];
+        uint256 stake = game.stake;
+        require(game.step == Steps.PLAYER_ONE_REVEAL, INVALID_STEP_MSG);
+        address playerTwo = game.playerTwo;
+        require(msg.sender == playerTwo, INVALID_PLAYER_MSG);
+        require(playerTwoMove < 3, INVALID_MOVE_MSG);
+
+        //Validate move
+        bytes32 expectedMoveHash = createOpponentMoveHash(gameKey, secret, playerTwoMove);
+        require(game.playerTwoMoveHash == expectedMoveHash, "Move and secret do not match");
+
+        uint8 playerOneMove = game.playerOneMove;
```

Comment:
Mmh, so `Rock == 0`? That makes it somewhat difficult to differentiate from _absence of data_.

Reply:
Yes. This was my thought also but the algorithm dictates this so a trade-off. Absence of data should be impossible because before:

` require(game.step == uint8(Steps.PLAYER_ONE_REVEAL), INVALID_STEP_MSG);`

game.playerOneMove must have been set by this stage.